### PR TITLE
[SPARK-19112][CORE] Support for ZStandard codec

### PR DESCRIPTION
**Pull request by sitalkedia**
## What changes were proposed in this pull request?

Using zstd compression for Spark jobs spilling 100s of TBs of data, we could reduce the amount of data written to disk by as much as 50%. This translates to significant latency gain because of reduced disk io operations. There is a degradation CPU time by 2 - 5% because of zstd compression overhead, but for jobs which are bottlenecked by disk IO, this hit can be taken. 

## Benchmark 
Please note that this benchmark is using real world compute heavy production workload spilling TBs of data to disk

|         | zstd performance as compred to LZ4   |
| ------------- | -----:|
| spill/shuffle bytes    | -48% |
| cpu time    |    + 3% | 
| cpu reservation time       |    -40%|
| latency     |     -40% |

  
## How was this patch tested?

Tested by running few jobs spilling large amount of data on the cluster and amount of intermediate data written to disk reduced by as much as 50%.

### CODE REVIEW

**Comment by sitalkedia** [comment]
cc - @srowen, @tgravescs, @rxin, @sameeragarwal

**Comment by sitalkedia** [comment]
Old PR - https://github.com/apache/spark/pull/17303

**Comment by rxin** [comment]
Any benchmark data?

**Comment by HyukjinKwon** [comment]
cc @dongjinleekr too.

**Comment by sitalkedia** [comment]
@rxin - Updated with benchmark data on our production workload.

**Comment by sitalkedia** [comment]
Please note that few minor improvements I have made when comapring to old PR - #17303
1. Use zstd compression level 1 instead of 3, which is significantly faster.
2. Wrap the zstd input/output stream in buffered input/output stream to avoid overhead of excessive JNI call.

**Comment by sitalkedia** [comment]
jenkins retest this please.

**Comment by sitalkedia** [comment]
Any idea what is the build failure about?

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala [hunk 1]
```diff
@@ -50,13 +51,14 @@ private[spark] object CompressionCodec {
...
-      || codec.isInstanceOf[LZ4CompressionCodec])
+      || codec.isInstanceOf[LZ4CompressionCodec] || codec.isInstanceOf[ZStandardCompressionCodec])
   }
 
   private val shortCompressionCodecNames = Map(
     "lz4" -> classOf[LZ4CompressionCodec].getName,
     "lzf" -> classOf[LZFCompressionCodec].getName,
-    "snappy" -> classOf[SnappyCompressionCodec].getName)
+    "snappy" -> classOf[SnappyCompressionCodec].getName,
+    "zstd" -> classOf[SnappyCompressionCodec].getName)
```
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: you mean `ZStandardCompressionCodec` ?

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala [hunk 2]
```diff
@@ -216,3 +218,30 @@ private final class SnappyOutputStreamWrapper(os: SnappyOutputStream) extends Ou
     }
   }
 }
+
+/**
+ * :: DeveloperApi ::
+ * ZStandard implementation of [[org.apache.spark.io.CompressionCodec]].
```
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: would be good to add this link to the spec : http://facebook.github.io/zstd/

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala [hunk 3]
```diff
@@ -216,3 +218,30 @@ private final class SnappyOutputStreamWrapper(os: SnappyOutputStream) extends Ou
...
+ *
+ * @note The wire protocol for this codec is not guaranteed to be compatible across versions
+ * of Spark. This is intended for use as an internal compression utility within a single Spark
+ * application.
+ */
+@DeveloperApi
+class ZStandardCompressionCodec(conf: SparkConf) extends CompressionCodec {
+
+  override def compressedOutputStream(s: OutputStream): OutputStream = {
+    val level = conf.getSizeAsBytes("spark.io.compression.zstandard.level", "1").toInt
```
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: please add a comment explaining the reason why we chose level 1 over other levels

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala [hunk 4]
```diff
@@ -216,3 +218,30 @@ private final class SnappyOutputStreamWrapper(os: SnappyOutputStream) extends Ou
...
+    val compressionBuffer = conf.getSizeAsBytes("spark.io.compression.lz4.blockSize", "32k").toInt
```
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: - wondering if we should share this config value OR have a new one.
- do you want to set the default to something higher like 1mb or 4mb ?

**Comment by tejasapatil** [comment]
In `Benchmark` section the values for `Lz4` are all zeros which feels confusing while reading.. first thing I thought is they were absolute values but they are supposed to be relative

**Comment by tejasapatil** [comment]
re build failure: you can repro that locally by running "./dev/test-dependencies.sh". Its failing due to introducing a new dep... you need to add it to `dev/deps/spark-deps-hadoop-XXX`

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala (hunk 4 above)
Comment (sitalkedia) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: You are right, we should not share the config with lz4, created a new one.
Lets keep the default to 32kb which is aligned with the block size used by other compressions.

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala (hunk 1 above)
Comment (sitalkedia) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: Ah, my bad. Fixed it.

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala (hunk 2 above)
Comment (sitalkedia) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: done.

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala (hunk 3 above)
Comment (sitalkedia) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: done.

**Comment by rxin** [comment]
How big is the dependency that's getting pulled in? If we are adding more compression codecs maybe we should retire some old ones, or move them into a separate package so downstream apps can optionally depend on them.

**Comment by srowen** [comment]
Why does this need to be in Spark? and what are the licensing terms of the native code underneath (just suspicious because it's often GPL)? can a user not just add this with their app?

I tend to think we support what Hadoop supports for us here. Doesn't a later Hadoop pull this in?

**Comment by tgravescs** [comment]
> Why does this need to be in Spark? 
@srowen  you already asked that question and it has been answered on the jira as well as the old pr.  A user cannot add zstd compression to the internal spark parts:   spark.io.compression.codec. In this particular case he is saying its the shuffle output where its making a big difference.
zstd is already included in other open source projects like Hadoop, but again we don't get that for Spark internal compression code, zstd itself is BSD license. It looks like this pr is using the https://github.com/luben/zstd-jni which also appears to be BSD licensed.   We need to decide if using that is ok for us to use directly.  Hadoop wrote its own version but I would say if that version is working we use it.  Worse case if something happens where that user won't fix something we could fork it and aren't any worse then having our own copy to start with.

**Comment by srowen** [comment]
Got it, thanks for the reminder. I think the question is mostly about license and dependency weight then. I think we'd want to use whatever Hadoop provides.

File: docs/configuration.md [hunk 5]
```diff
@@ -866,7 +866,8 @@ Apart from these, the following properties are also available, and may be useful
     e.g.
     <code>org.apache.spark.io.LZ4CompressionCodec</code>,
     <code>org.apache.spark.io.LZFCompressionCodec</code>,
-    and <code>org.apache.spark.io.SnappyCompressionCodec</code>.
+    <code>org.apache.spark.io.SnappyCompressionCodec</code>.
```
Comment (kiszk) [inline docs/configuration.md]: nit: '</code>.' -> '</code>,'

File: docs/configuration.md [hunk 6]
```diff
@@ -886,6 +887,23 @@ Apart from these, the following properties are also available, and may be useful
   </td>
 </tr>
 <tr>
+  <td><code>spark.io.compression.zstd.level</code></td>
+  <td>1</td>
+  <td>
+    Compression leve for Zstd compression codec. Increasing the compression level will result in better
```
Comment (kiszk) [inline docs/configuration.md]: nit: leve -> level

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala [hunk 7]
```diff
@@ -216,3 +218,33 @@ private final class SnappyOutputStreamWrapper(os: SnappyOutputStream) extends Ou
...
+ * application.
+ */
+@DeveloperApi
+class ZStdCompressionCodec(conf: SparkConf) extends CompressionCodec {
+
+  override def compressedOutputStream(s: OutputStream): OutputStream = {
+    // Default compression level for zstd compression to 1 because it is
+    // fastest of all with reasonably high compression ratio.
+    val level = conf.getSizeAsBytes("spark.io.compression.zstd.level", "1").toInt
+    val bufferSize = conf.getSizeAsBytes("spark.io.compression.zstd.bufferSize", "32k").toInt
```
Comment (kiszk) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: Would it be better to have this variable as a private variable to get this property only once?

**Comment by rxin** [comment]
Our compression codec is actually completely decoupled from Hadoops, but dependency management (and licensing) can be annoying to deal with.

**Comment by sitalkedia** [comment]
>> How big is the dependency that's getting pulled in?

zstd-jni library actually is a very thin library and is not pulling any dependency of its own, so I would not be worried about dependency management issues. 

Let me know if we are fine using https://github.com/luben/zstd-jni library which uses BSD license. Otherwise, we might have to implement something of our own.

**Comment by vanzin** [comment]
> Let me know if we are fine using https://github.com/luben/zstd-jni library which uses BSD license. 

Isn't the license of the native library (not the JNI wrapper) the same or very similar to the RocksDB license that was recently blacklisted by the ASF? That would make it hard to add it to Spark.

(RocksDB has since moved to ASLv2.)

**Comment by tgravescs** [comment]
https://www.apache.org/legal/resolved.html

Seems to list bsd-2 and bsd-3 as ok. https://github.com/facebook/zstd/blob/dev/LICENSE  
not sure how the Facebook BSD+Patents license was showing up so we need to make sure it doesn't fall into the "Facebook BSD+Patents license" category

**Comment by tgravescs** [comment]
oh it might actually fall into that as it has the patents:
https://github.com/facebook/zstd/blob/dev/PATENTS

**Comment by vanzin** [comment]
Yes, the native library has the BSD+patents license:
https://github.com/luben/zstd-jni/tree/master/src/main/native

**Comment by rxin** [comment]
@sitalkedia anyway you can talk to the FB team that does that one and relicense, similar to RocksDB?

**Comment by sitalkedia** [comment]
@rxin - Sure, let me talk to folks internally to see if it is possible to relicense. Otherwise, we might have to upgrade to hadoop 2.9.0, which will come with its own zstd implementation.

**Comment by discipleforteen** [comment]
how about TPCDS ?

**Comment by sitalkedia** [comment]
Just an update on this - I am in talk with our internal team to relicense zstd library. This might take some time though. I will keep you updated.

@discipleforteen - Unfortunately, we do not have setup to run TPDCS benchmarks.  Feel free to run the benchmark on this change and report the number if you have the setup readily available.

**Comment by sitalkedia** [comment]
https://github.com/facebook/zstd/issues/775

**Comment by sitalkedia** [comment]
We have released new zstd version (https://github.com/facebook/zstd/releases) with modified BSD + GPLv2 license. 

@rxin, @srowen, @markhamstra - Can you confirm this looks fine from licensing perspective and this PR is good to go?

**Comment by sitalkedia** [comment]
Relevant PR - https://github.com/facebook/zstd/pull/801/files

**Comment by dongjoon-hyun** [comment]
Hi, @sitalkedia .

facebook/zstd#775 was `Consider re-licensing to Apache License v2`.

But, I found that the new one is not Apache License v2. It's about adding GPLv2. I'm wondering if  GPLv2 is enough for us.

For [Apache licensing](https://www.apache.org/licenses/GPL-compatibility.html), I found the following.
> Despite our best efforts, the FSF has never considered the Apache License to be compatible with GPL version 2

**Comment by vanzin** [comment]
The patents grant (`PATENTS` file) seems to have been dropped from the repo:
https://github.com/facebook/zstd

That makes it plain BSD, which is fine for Spark.
(https://github.com/facebook/zstd/blob/dev/LICENSE)

**Comment by dongjoon-hyun** [comment]
Thank you for confirming, @vanzin ! I see.

**Comment by markhamstra** [comment]
In addition to LICENSE, there is also COPYING in the v1.3.1 release:

https://github.com/facebook/zstd/blob/v1.3.1/LICENSE
https://github.com/facebook/zstd/blob/v1.3.1/COPYING

I'm not going to pretend to have a fully informed opinion on this issue.

**Comment by srowen** [comment]
It's dual-licensed (see bottom of https://github.com/facebook/zstd/blob/v1.3.1/README.md ) and one is 3-clause BSD. This is fine as it's a "Category A" license: https://www.apache.org/legal/resolved.html  

The zstd-jni binding itself is 2-clause BSD, also OK. https://github.com/luben/zstd-jni/blob/master/LICENSE

Licenses check out and there's no extra dependencies it drags in.

I think this will be OK but we do need to add these two licenses to `licenses/` (see the convention there) and also add a line for each in `LICENSE` here.

**Comment by vanzin** [comment]
Just to make sure, this is ok if the JNI wrapper does not bring the zstd native library with it; otherwise it also needs to be updated since it still has the old BSD+PATENTS licence (https://github.com/luben/zstd-jni/tree/master/src/main/native), and Spark needs to pull in the version with the correct license.

**Comment by sitalkedia** [comment]
>> I think this will be OK but we do need to add these two licenses to licenses/ (see the convention there) and also add a line for each in LICENSE here.

@srowen - Does that need to be done with this PR? What are the next steps for this PR?

**Comment by srowen** [comment]
Yes, licenses have to be updated, that's the one type of thing that's not optional.
But Marcelo is right that the library actually doesn't yet include the newer dependency with the right license. We can't pull it in until it pulls in 1.3.1.

**Comment by tgravescs** [comment]
It looks like zstd-jni has now been updated to pull 1.3.1

**Comment by sitalkedia** [comment]
Updated with zstd-jni versin 1.3.1-1 and also updated the license to include zstd-jni license. @srowen - How does that look from licensing prospective?

File: dev/deps/spark-deps-hadoop-2.6 [hunk 8]
```diff
@@ -186,3 +186,4 @@ xercesImpl-2.9.1.jar
 xmlenc-0.52.jar
 xz-1.0.jar
 zookeeper-3.4.6.jar
+zstd-jni-1.3.0-1.jar
```
Comment (srowen) [inline dev/deps/spark-deps-hadoop-2.6]: These need to be updated, it seems

File: LICENSE [hunk 9]
```diff
@@ -269,6 +269,7 @@ The text of each license is also included at licenses/LICENSE-[project].txt.
      (BSD 3 Clause) d3.min.js (https://github.com/mbostock/d3/blob/master/LICENSE)
      (BSD 3 Clause) DPark (https://github.com/douban/dpark/blob/master/LICENSE)
      (BSD 3 Clause) CloudPickle (https://github.com/cloudpipe/cloudpickle/blob/master/LICENSE)
+     (BSD 2 Clause) Zstd-jni (https://github.com/luben/zstd-jni/blob/master/LICENSE)
```
Comment (srowen) [inline LICENSE]: ZStandard itself is also included, not just zstd-jni. We'd need an entry for it, and its LICENSE as below.
https://github.com/luben/zstd-jni/tree/master/src/main/native

**Review by srowen — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala (hunk 7 above)
Comment (srowen) [reply core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: Agree, it's simpler and cleaner, as it avoids duplicating this property in this file

**Comment by srowen** [comment]
Hm, this might be a real error. It seems to be hanging at:

```
[info] ExternalAppendOnlyMapSuite:
...
[info] - simple cogroup (56 milliseconds)
[info] - spilling (3 seconds, 678 milliseconds)
9/21/17 1:43:14 AM =============================================================

-- Gauges ----------------------------------------------------------------------
master.aliveWorkers
9/21/17 2:03:14 AM =============================================================

-- Gauges ----------------------------------------------------------------------
master.aliveWorkers
9/21/17 2:23:14 AM =============================================================
...
```

However I can't figure out how that would be related. This new codec isn't enabled anywhere except its test. Hm, I can run it again tomorrow in case it something else transient, but it did fail this way twice and nothing else did.

**Comment by sitalkedia** [comment]
Thanks for looking into this @srowen. Its weird, I dont understand that either. Also, I am not able to reproduce this issue on my laptop.

**Comment by sitalkedia** [comment]
@srowen - Did you get a chance to look into this ?

**Comment by srowen** [comment]
I didn't look into it. I can restart the test here though

**Comment by vanzin** [comment]
retest this please

**Comment by sitalkedia** [comment]
@vanzin - May be the test time outs are related to one test failure -https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82608/testReport/org.apache.spark.scheduler/ReplayListenerSuite/_It_is_not_a_test_it_is_a_sbt_testing_SuiteSelector_/

Its failing to load the native library -

```building from source the jar or providing libzstd-jni in you system.
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1864)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at com.github.luben.zstd.util.Native.load(Native.java:101)
	at com.github.luben.zstd.ZstdOutputStream.<clinit>(ZstdOutputStream.java:17)
	at org.apache.spark.io.ZStdCompressionCodec.compressedOutputStream(CompressionCodec.scala:244)
	at org.apache.spark.scheduler.EventLoggingListener$$anonfun$2.apply(EventLoggingListener.scala:122)
	at org.apache.spark.scheduler.EventLoggingListener$$anonfun$2.apply(EventLoggingListener.scala:122)
	at scala.Option.map(Option.scala:146)
	at org.apache.spark.scheduler.EventLoggingListener.start(EventLoggingListener.scala:122)
	at org.apache.spark.SparkContext.<init>(SparkContext.scala:524)
	at org.apache.spark.SparkContext.<init>(SparkContext.scala:127)
	at ```
 
Unfortunately, I am not able to repro this issue on my laptop. Any idea what might be wrong?

**Comment by rxin** [comment]
Does the package include a binary distribution for Linux?

**Comment by sitalkedia** [comment]
Yes, the binary distribution is included in the zstd-jni jar file.

**Comment by vanzin** [comment]
The [code](https://github.com/luben/zstd-jni/blob/master/src/main/java/com/github/luben/zstd/util/Native.java) overwrites the original exception message that might shed some light on what's going on... and also ignores some exceptions it shouldn't be ignoring (like errors on close, which may indicate low disk space).

Anyway, let's try again to see if it's at least consistent.

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
Same test failed, so looks like there's a real non-infra-related issue...

**Comment by hvanhovell** [comment]
retest this please

**Comment by vanzin** [comment]
I haven't been able to reproduce the issue locally, but looking at the jenkins logs I see a bunch of exceptions like these:

```
17/10/13 06:53:26.609 dispatcher-event-loop-15 ERROR Worker: Failed to launch executor app-20171013030138-0000/3 for Test replay.
java.lang.IllegalStateException: Shutdown hooks cannot be modified during shutdown.
        at org.apache.spark.util.SparkShutdownHookManager.add(ShutdownHookManager.scala:195)
```

And:

```
17/10/13 06:53:26.687 pool-1-thread-1-ScalaTest-running-ExternalAppendOnlyMapSuite WARN SparkContext: Another SparkContext is being constructed (or threw an exception in its constructor).  This may indicate an error, since only one SparkContext may be running in this JVM (see SPARK-2243). The other SparkContext was created at:
org.apache.spark.SparkContext.<init>(SparkContext.scala:127)
org.apache.spark.util.collection.ExternalAppendOnlyMapSuite$$anonfun$12.apply$mcV$sp(ExternalAppendOnlyMapSuite.scala:30
```

Note that the first error mentions the app name used by `ReplayListenerSuite` but it actually happens in a completely separate test suite. At the very least, `ReplayListenerSuite` is doing a poor job of cleaning up after itself and we should fix that.

**Comment by vanzin** [comment]
(I'll file a bug and send a PR for it separately, btw.)

**Comment by vanzin** [comment]
Turns out that's caused by SparkContext failing to clean up after itself when the `UnsatisfiedLinkError` happens, so those errors are red herrings...

**Comment by hvanhovell** [comment]
This seems to be caused by a issue in the `zstd-jni` library. It probably uses the wrong `ClassLoader` to load the native library, and as a result it cannot find the library & load it.

**Comment by vanzin** [comment]
Yeah but that would also cause it to fail locally if it were the cause, and it passes for me. I can't really figure out from the rest of the logs if something obvious is wrong, so I guess the best bet now is to as for changes in the `zstd-jni` so that all errors are properly reported (see https://github.com/apache/spark/pull/18805#issuecomment-335885053).

**Comment by vanzin** [comment]
Good news is that I can reproduce it on the amplab machine, so I'll try to play around with the zstd-jni code a bit.

**Comment by vanzin** [comment]
```
$ ldd linux/amd64/libzstd-jni.so 
ldd: warning: you do not have execution permission for `linux/amd64/libzstd-jni.so'
linux/amd64/libzstd-jni.so: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by linux/amd64/libzstd-jni.so)
        linux-vdso.so.1 =>  (0x00007ffe0dfda000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f89eb3a4000)
        /lib64/ld-linux-x86-64.so.2 (0x0000003612e00000)
```

Mystery solved; library is compiled with a newer glibc requirement than the amplab machines have. Can we ask them to tweak their compilation to support older Linux distros?

```
$ cat /etc/issue
CentOS release 6.9 (Final)
```

**Comment by sitalkedia** [comment]
Created https://github.com/luben/zstd-jni/issues/47.

**Comment by sitalkedia** [comment]
ping.

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala (hunk 7 above)
Comment (srowen) [reply core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: @sitalkedia how about comments like this?

**Review by srowen — COMMENTED** [review]
Aside from one minor question from an old comment that's looking good. The licenses seem in order.

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala (hunk 7 above)
Comment (sitalkedia) [reply core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: Sorry somehow missed these comments. Will address.

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala [hunk 10]
```diff
@@ -216,3 +218,33 @@ private final class SnappyOutputStreamWrapper(os: SnappyOutputStream) extends Ou
...
+ * of Spark. This is intended for use as an internal compression utility within a single Spark
+ * application.
+ */
+@DeveloperApi
+class ZStdCompressionCodec(conf: SparkConf) extends CompressionCodec {
+
+  override def compressedOutputStream(s: OutputStream): OutputStream = {
+    // Default compression level for zstd compression to 1 because it is
+    // fastest of all with reasonably high compression ratio.
+    val level = conf.getSizeAsBytes("spark.io.compression.zstd.level", "1").toInt
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: Is this `getInt` instead of `getSizeAsBytes`?

Comment (sitalkedia) [reply core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: Good eye, fixed.

File: core/src/main/scala/org/apache/spark/io/CompressionCodec.scala [hunk 11]
```diff
@@ -216,3 +218,33 @@ private final class SnappyOutputStreamWrapper(os: SnappyOutputStream) extends Ou
...
+ * details see - http://facebook.github.io/zstd/
+ *
+ * @note The wire protocol for this codec is not guaranteed to be compatible across versions
+ * of Spark. This is intended for use as an internal compression utility within a single Spark
+ * application.
+ */
+@DeveloperApi
+class ZStdCompressionCodec(conf: SparkConf) extends CompressionCodec {
+
+  val bufferSize = conf.getSizeAsBytes("spark.io.compression.zstd.bufferSize", "32k").toInt
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: This should be private. The intent was to lift both config values out of the method, so level can do here too.

Comment (sitalkedia) [reply core/src/main/scala/org/apache/spark/io/CompressionCodec.scala]: done.

**Review by srowen — APPROVED** [review]

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by hvanhovell** [comment]
Merging to master. Thanks for seeing this through!

**Pull request closed without merging**